app.go

```diff
@@ -475,6 +506,39 @@ func New(opts ...Option) *App {
 		app.provide(p)
 	}
 
+	// If WithLogger and Printer are both provided, WithLogger takes precedence.
+	// If user supplied an fxlog constructor, we need to have dig's container
+	// initialize it and grab all of arguments from the graph and then return.
+	if app.fxlogConstructor != nil {
+		err := app.container.Provide(app.fxlogConstructor)
+		if err != nil {
+			err = fmt.Errorf("could not construct customer logger via fx.WithLogger: %w", err)
+			app.err = err
+
+			return app
+		}
+		var userLogger fxlog.Logger
+		if err := app.container.Invoke(func(got fxlog.Logger) { userLogger = got }); err != nil {
+			fxlog.Error("error constructing logger, did not get back fxlog.Logger").Write(app.logBuffer)
+			app.err = err
+
+			return app
+		}
+
+		app.log = userLogger
+		fxlog.Info("setting custom fxlog.Logger",
+			fxlog.F("constructor", fxreflect.FuncName(app.fxlogConstructor))).Write(app.logBuffer)
+	}
+	// Initialize a logger if user hasn't provided one via WithLogger.
+	if app.log == nil {
+		app.log = fxlog.DefaultLogger(os.Stderr)
+	}
+	// Flush out all previously buffered logs and connect newly initialized logger.
+	app.logBuffer.Connect(app.log)
```

Comment:
We probably want the logBuffer.Connect as suggested above so that it happens after all the other Provides.

That way, the log constructor can actually have a dependency on the fx.Lifecycle. (Which makes sense in case the logger has some kind of Shutdown functionality.)